app/src/main/java/com/example/applemusicanimation/ui/main/MainActivity.kt

```diff
@@ -19,27 +22,78 @@ class MainActivity : BaseActivity<ActivityMainBinding, MainViewModel>() {
         super.onCreate(savedInstanceState)
         onClickListener()
         addObservers()
+        setUpSeekBar()
     }
 
-    private fun addObservers(){
-        viewModel.playWhenReady.observe(this, Observer {isPlaying->
-            binding.pauseButton.setOnClickListener {
-                viewModel.playPauseToggle(isPlaying)
-            }
-        })
+    private fun addObservers() {
+
+        setSleepToggle()
     }
 
     @SuppressLint("ClickableViewAccessibility")
     private fun onClickListener() {
+
         with(binding) {
             first.setOnTouchListener { _, event ->
-                if (event.action == MotionEvent.ACTION_UP){
-                    viewModel.initialisePlayer()
+                if (event.action == MotionEvent.ACTION_UP) {
+                    viewModel.setupPlayer(0)
                     Timber.d("clicked")
                 }
                 return@setOnTouchListener false
             }
+            nextButton.setOnClickListener { viewModel.nextSong() }
+
+            previousButton.setOnClickListener { viewModel.previousSong() }
+
+            seekBar.setOnSeekBarChangeListener(this@MainActivity)
 
+            binding.pauseButton.setOnClickListener { viewModel.playPauseToggle() }
         }
     }
+
+    private fun setSleepToggle() {
+        var isSleepMode = false
+
+        viewModel.isSleepMode.observe(this, Observer {
+            with(binding.sleepButton) {
+                isSleepMode = it
+                if (it)
+                    setBackgroundResource(
+                        R.drawable.ic_timer_off_black_24dp
+                    )
+                else
+                    setBackgroundResource(
+                        R.drawable.ic_timer_black_24dp
+                    )
+            }
+        })
+        sleepListener(isSleepMode)
+    }
+
+    private fun sleepListener(isSleepMode: Boolean) {
+        binding.sleepButton.setOnClickListener {
+            if (!isSleepMode)
+                viewModel.startCountDownTimer(30000)
+            else
+                viewModel.stopCountDownTimer()
```

Comment:
This logic can be moved to `viewModel`. We can have a function in viewModel called `toggleSleepMode()`.
ViewModel already has the state i.e. `isSleepMode`. Based on it, `startCountdownTimer()` and `stopCountDownTimer()` can be called.

This way `startCountDownTimer()` and `stopCountDownTimer()` can be private to viewModel.